compiler/rustc_hir_analysis/src/astconv/lint.rs

```diff
@@ -94,7 +97,11 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         let is_object_safe = match self_ty.kind {
             hir::TyKind::TraitObject(objects, ..) => {
                 objects.iter().all(|o| match o.trait_ref.path.res {
-                    Res::Def(DefKind::Trait, id) => tcx.check_is_object_safe(id),
+                    Res::Def(DefKind::Trait, id) => {
+                        // When we're dealing with a recursive trait, we don't want to downgrade
+                        // the error, so we consider them to be object safe always. (#119652)
+                        tcx.check_is_object_safe(id) || Some(id) == owner
```

Comment:
Can this be swapped to avoid executing the query if not necessary or would that change the meaning?

Reply:
You mean something like the following?

```rust
                    Res::Def(DefKind::Trait, id) if Some(id) == owner => {
                        // When we're dealing with a recursive trait, we don't want to downgrade
                        // the error, so we consider them to be object safe always. (#119652)
                        true
                    }
                    Res::Def(DefKind::Trait, id) => {
                        tcx.check_is_object_safe(id)
                    }
```

Reply:
This seems to be causing some issues for me generating accurate lint suggestions, because types that are not actually object safe are now being marked as object safe?

Reply:
@trevyn We can also entirely remove the downgrade, given how many head aches this is giving us, but your PR looks correct.